reco_utils/recommender/deeprec/io/nextitnet_iterator.py

```diff
@@ -0,0 +1,269 @@
+                # label_list_all.append(1)
+                item_list_all.append(positive_item)
+                item_cate_list_all.append(positive_item_cate)
+
+                count = 0
+                while count < batch_num_ngs:
+                    negative_item_list = []
+                    negative_item_cate_list = []
+                    count_inner = 1
+                    while count_inner <= max_seq_length_batch:
+                        random_value = random.randint(0, instance_cnt - 1)
+                        negative_item = item_list[random_value]
+                        if negative_item == positive_item[count_inner - 1]:
+                            continue
+                        negative_item_list.append(negative_item)
+                        negative_item_cate_list.append(item_cate_list[random_value])
+                        count_inner += 1
+
+                    label_list_all.append([0] * max_seq_length_batch)
+                    # label_list_all.append(0)
+                    item_list_all.append(negative_item_list)
+                    item_cate_list_all.append(negative_item_cate_list)
+                    count += 1
+
+            res = {}
+            res["labels"] = np.asarray(
+                label_list_all, dtype=np.float32
+            )  # .reshape(-1,1)
+            res["users"] = user_list_all
+            res["items"] = np.asarray(item_list_all, dtype=np.int32)
+            res["cates"] = np.asarray(item_cate_list_all, dtype=np.int32)
+            res["item_history"] = item_history_batch_all
+            res["item_cate_history"] = item_cate_history_batch_all
+            res["mask"] = mask
+            res["time"] = time_list_all
+            res["time_diff"] = time_diff_batch
+            res["time_from_first_action"] = time_from_first_action_batch
+            res["time_to_now"] = time_to_now_batch
+
+            # print("label_list_all.shape: ", res["labels"].shape)
+            # print("item_list_all.shape: ", res["items"].shape)
+
+            return res
+
+        else:
+            instance_cnt = len(label_list)
+            history_lengths = [len(item_history_batch[i]) for i in range(instance_cnt)]
+            max_seq_length_batch = self.max_seq_length
+            item_history_batch_all = np.zeros(
+                (instance_cnt, max_seq_length_batch)
+            ).astype("int32")
+            item_cate_history_batch_all = np.zeros(
+                (instance_cnt, max_seq_length_batch)
+            ).astype("int32")
+            time_diff_batch = np.zeros((instance_cnt, max_seq_length_batch)).astype(
+                "float32"
+            )
+            time_from_first_action_batch = np.zeros(
+                (instance_cnt, max_seq_length_batch)
+            ).astype("float32")
+            time_to_now_batch = np.zeros((instance_cnt, max_seq_length_batch)).astype(
+                "float32"
+            )
+            mask = np.zeros((instance_cnt, max_seq_length_batch)).astype("float32")
```

Comment:
All the `np.zeros(...).astype(..)` should be `np.zeros(..., dtype=...)` to avoid unnecessary array creation.

Reply:
good catch, changed